File: day-3/part-1/elixir/power_consumption.exs
defmodule PowerConsumption do

  def calculate(input) do
    report = prepare_report(input)
    report
      |> Enum.map(fn row ->
        row
          |> Enum.min_max_by(fn x ->
            Enum.count(row, fn y -> y == x end)
          end)
      end)
      |> Enum.reduce({"", ""}, fn {min, max}, {first, last} ->
        {first <> min, last <> max}
      end)
      |> Tuple.to_list()
      |> Enum.map(fn binary ->
        {number, _} = Integer.parse(binary, 2)
        number
      end)
      |> List.to_tuple()
      |> Tuple.product()
  end

  defp prepare_report(input) when is_binary(input) do
    input
      |> String.split("\n")
      |> Enum.map(&String.trim/1)
      |> Enum.filter(&(String.length(&1) > 0))
      |> Enum.map(&String.split(&1, "", trim: true))
      |> transpose()
  end

  defp prepare_report(file) do
    case file do
      {:ok, content} -> prepare_report(content)
      {:error, _} -> []
    end
  end

  defp transpose(rows) do
    rows
    |> List.zip
    |> Enum.map(&Tuple.to_list/1)
  end
end

File: day-3/part-1/ruby/power_consumption.rb
require 'matrix'

class PowerConsumption
  def calculate(input)
    report = prepare_report(input)
    least_most = report.map do |row|
      row.tally.minmax_by { |_k, v| v}.to_h.keys
    end
    least_most.reduce(["", ""]) do |acc, (min, max)|
      acc[0] << min
      acc[1] << max
      acc
    end.map { _1.to_i(2) }.reduce(1, :*)
  end

  private

  def prepare_report(input)
    report = input.split("\n").map { _1.split('') }
    Matrix[*report].transpose.to_a
  end
end